### PR TITLE
feat(npm): fuzzy merge registries in .yarnrc.yml

### PR DESCRIPTION
**Pull request by RGunning**
## Changes

match registries in the npmRegistries fuzzily, so that //my.registry will match with https://my.registry when merging .yarnrc.yml files instead of creating a .yarnrc.yml with 2 registries.

## Context

see: #26919 

At the moment, when constructing the yarnrc from the hostRules the https: prefix is stripped off the url, however, configuring yarn@4 with `yarn config set --home npmRegistryServer "https://my-private-regsistry"  && yarn config login` would set the protocol on the registry on the developers machine. this then causes problems when a repository needs to set additional npm config e.g. npmAlwaysAuth

Proposed fix is that instead of merging in such a way that 2 items in the npmRegistries are created, we instead fizzy match to identify that the registry is the same so they should be the same item in the registries array.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/post-update/index.spec.ts [hunk 1]
```diff
@@ -22,6 +23,7 @@ jest.mock('../../../../util/git');
 jest.mock('./npm');
 jest.mock('./yarn');
 jest.mock('./pnpm');
+jest.mock('./rules');
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/index.spec.ts:26]: you don't need to mock if you use a spy
```suggestion
```

File: lib/modules/manager/npm/post-update/index.ts [hunk 2]
```diff
@@ -563,7 +563,6 @@ export async function getAdditionalFiles(
     await updateNpmrcContent(lockFileDir, npmrcContent, additionalNpmrcContent);
     let yarnRcYmlFilename: string | undefined;
     let existingYarnrcYmlContent: string | undefined | null;
-    // istanbul ignore if: needs test
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/index.ts:566]: thank you 🤗

File: lib/modules/manager/npm/post-update/index.ts [hunk 3]
```diff
@@ -573,10 +572,28 @@ export async function getAdditionalFiles(
...
+              .map(([k, v]) => {
+                const key =
+                  Object.keys(existingYarnrRcYml.npmRegistries || {}).find(
+                    // match without trailing slashes
+                    (x) => x.replace(/\/$/, '').endsWith(k.replace(/\/$/, '')),
+                  ) ?? k;
+                return { [key]: v };
+              })
+              .reduce((acc, cur) => ({ ...acc, ...cur }), {}),
+          };
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/index.ts:590]: please move to separate function in module scope

**Review by viceice — COMMENTED** [review]
looks good, just some small things 😊

File: lib/modules/manager/npm/post-update/index.ts [hunk 4]
```diff
@@ -322,6 +322,24 @@ async function resetNpmrcContent(
...
+>(additionalYarnRcYml: T, existingYarnrRcYml: T): T {
+  return {
+    ...additionalYarnRcYml,
+    npmRegistries: Object.entries(additionalYarnRcYml.npmRegistries ?? {})
+      .map(([k, v]) => {
+        const key =
+          Object.keys(existingYarnrRcYml.npmRegistries ?? {}).find(
+            // match without trailing slashes
+            (x) => x.replace(/\/$/, '').endsWith(k.replace(/\/$/, '')),
+          ) ?? k;
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/index.ts:336]:
```suggestion
    keys.get(k) ?? k;
```


create a map outside the map loop, so we do this once

File: lib/modules/manager/npm/post-update/index.ts [hunk 5]
```diff
@@ -322,6 +322,24 @@ async function resetNpmrcContent(
   }
 }
 
+export function fuzzyMatchAdditionalYarnrcYml<
+  T extends { npmRegistries?: Record<string, unknown> },
+>(additionalYarnRcYml: T, existingYarnrRcYml: T): T {
+  return {
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/index.ts:328]:
```suggestion
  const keys = new Map(Object.keys(existingYarnrRcYml.npmRegistries ?? {}).map(
            (x) => [x.replace(/\/$/, '').replace(/^https?:/, ''), x]
          );
  return {
```
something like this

File: lib/modules/manager/npm/post-update/index.ts [hunk 6]
```diff
@@ -573,13 +593,18 @@ export async function getAdditionalFiles(
...
+            fuzzyMatchAdditionalYarnrcYml(
+              additionalYarnRcYml,
+              existingYarnrRcYml,
+            ),
           );
+
           await writeLocalFile(yarnRcYmlFilename, dump(updatedYarnYrcYml));
           logger.debug('Added authentication to .yarnrc.yml');
-        } catch (err) {
+        } catch (err) /* istanbul ignore next */ {
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/index.ts:607]: No, why this isn't hit anymore?

File: lib/modules/manager/npm/post-update/index.ts [hunk 7]
```diff
@@ -322,6 +322,27 @@ async function resetNpmrcContent(
   }
 }
 
+export function fuzzyMatchAdditionalYarnrcYml<
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/index.ts:325]: sorry, but one last thing. this new function should be moved to `./yarn.ts`.

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/npm/post-update/index.ts (hunk 7 above)
Comment (RGunning) [reply lib/modules/manager/npm/post-update/index.ts]: Done, but as index.spec.ts mocks the yarn.ts import I removed some of the tests from index.spec.ts. Coverage is still there, but tests no longer verify the resultant yarnrc file to the same extent

**Review by viceice — APPROVED** [review]

**Pull request merged**